Environment/Environment.tex

```diff
@@ -620,27 +628,53 @@ \subsection{Packages: declarative categorization of Squeak code}
 \end{code}
 
 \noindent
-These methods belong in \damien{don't we say 'belong to'?} the same package as the classes in the categories \ct{SqueakLink-Connections} and \ct{SqueakLink-Model}.
+These methods belong in the same package as the classes in the categories \ct{SqueakLink-Connections} and \ct{SqueakLink-Model}.
 But clearly the whole of class \ct{Object} does not belong in your package!
 So you need a way of putting certain \emph{methods} in a package, even though the rest of the class is in another package.
 \index{package!extension}
-\seeindex{extension package}{package, extension}
+\seeindex{extension method}{package, extension}
+
+This is where the concept of \ind{extension methods} comes in.
+You can mark these converter methods as extension methods by placing them in a protocol (of \ct{Object}, \ct{String}, \ct{Date}, and so on) named \prot{*SqueakLink-Model-converting} (note the initial asterisk).
+% In general, any method in any protocol that starts with an asterisk, followed by a package name, followed by a hyphen, is an extension method to that package.
+
+\subsubsection{Override methods}
+
+Sometimes, you may also want to alter a method from another package that already exists and store the changed method in your package.
+For instance, this can be required if you need to extend an existing feature or fix a bug in the standard Squeak image.
+Let's say that you want to add a menu item to an inspector's context menu for opening its SQL version, so you have edited \ct{Inspector>>addObjectItemsTo:} and now would like to check it into your package repository.
+\index{package!extension!overriding}
+\seeindex{override method}{package, extension, overriding}
+
+While you could place it in a simple extension protocol as above, this would introduce some problems:
+First, the method would be displayed as missing in the original package, so the package would unintentionally become dirty.
+Second, if you unloaded your own package again, the original method would be lost.
+
+A better way to do this is to mark the method as an \ind{overriding method} by placing it in a protocol named \prot{*SqueakLink-menu-override}.
+In general, an override protocol is an extension protocol that ends with the suffix \prot{-override}.
+Usually, it's a good idea to repeat the name of the original protocol in the override protocol name.
+
+However, beware that every change to an existing package might break other packages that depend on it.
+Thus, use override methods as seldom as possible.
+Often, there are better ways to extend the functionality of a package, such as subclassing or using explicit hooks.\footnote{Actually, there is a better way to extend the menu in our example: Place a new (non-override) extension method named \ct{squeakLinkMenu:} in the \ct{Inspector} class and add the pragma \ct{<fieldListMenu>} to it.}

```

Comment:
"Actually, there is an even better way to extend the menu in our example:"

Reply:
```suggestion
Often, there are better ways to extend the functionality of a package, such as subclassing or using explicit hooks.\footnote{Actually, there is an even better way to extend the menu in our example: Place a new (non-override) extension method named \ct{squeakLinkMenu:} in the \ct{Inspector} class and add the pragma \ct{<fieldListMenu>} to it.}
```